### PR TITLE
Use ume

### LINKED ISSUE #46: Bart
I have a solution to fix this situation, and I will sent a pull request today.

### PR DESCRIPTION
**Pull request by ryo33**
Closes #46 

- [x] 1. Pass `cargo test`
- [x] 2. Get the go-ahead
- [x] ~~3. Separate code into separate modules~~
- [x] 4. Replace all `BartDisplay`
- [x] 5. Keep output the same as before
- [x] 6. Publish `ume` and use it

### CODE REVIEW

**Comment by ryo33** [comment]
I've used `ume` for the most complex and simple templates. It keeps output as before except for whitespaces, comments, and `use` (it's easy to fix).

We lost declarativity but got the expressiveness of Rust (and syntax highlights). I think the discussion "which is better?" could be bikeshedding, and I don't believe `ume` is better than `bart` (IMO `ume` is only slightly better, but there may be a bias in this). For this reason, I would like to leave the decision to @jsinger67.

**Comment by ryo33** [comment]
If accepted, I expect all tasks can be completed within a day or a few days.

**Comment by jsinger67** [comment]
In my opinion `ume` has the great advantage of being actively developed by a person with an own interest in `parol` 😎.
This isn't the case for `bart`. So I think it's strategically justified.
Your checklist shows me that you have a good plan for completing this task.
Give it a try and good luck from my side. 👍

**Comment by ryo33** [comment]
I've found [genco](https://github.com/udoprog/genco), which is richer than `ume`.
We can replace `ume` with `genco` in the future.
Contrary to `ume`, `genco` supports:
- comments handling
- whitespace handling

**Comment by jsinger67** [comment]
The generated code in `parol` is only semi-formatted and in a postprocess formatted with the help of [rustfmt](https://github.com/jsinger67/parol/blob/main/crates/parol/src/generators/rust_code_formatter.rs).
Therefore we need not to put a lot of energy into formatting during the generation phase.

File: crates/parol-ls/src/parol_ls_grammar_trait.rs [hunk 1]
```diff
@@ -11,14 +11,12 @@
...
-#[allow(unused_imports)]
-use parol_runtime::parol_macros::{pop_and_reverse_item, pop_item};
-
 use parol_runtime::log::trace;
 #[allow(unused_imports)]
 use parol_runtime::miette::{bail, miette, IntoDiagnostic, Result};
+#[allow(unused_imports)]
+use parol_runtime::parol_macros::{pop_and_reverse_item, pop_item};
 use parol_runtime::parser::{ParseTreeStackEntry, ParseTreeType, UserActionsTrait};
-
```
Comment (ryo33) [inline crates/parol-ls/src/parol_ls_grammar_trait.rs:21]: This change was inevitable. The output is identical for all crates except for `parol-ls`.

File: crates/parol-ls/src/parol_ls_grammar_trait.rs [hunk 2]
```diff
@@ -1149,8 +1147,7 @@ where
     // Mutable reference of the actual user grammar to be able to call the semantic actions on it
     user_grammar: &'u mut dyn ParolLsGrammarTrait,
     // Stack to construct the AST on it
-    item_stack: Vec<ASTType>,
-    // Just to hold the lifetime generated by parol
+    item_stack: Vec<ASTType>, // Just to hold the lifetime generated by parol
```
Comment (ryo33) [inline crates/parol-ls/src/parol_ls_grammar_trait.rs:1150]: I should fix this.

File: crates/parol/tests/parol_new.rs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        .current_dir(&path)
+        .args(["new", "--bin", "--path", "bin", "--name", "snapshot_bin"])
+        .assert()
+        .success();
+
+    diff(
+        path.path().join("bin").to_str().unwrap(),
+        snapshot_path("bin"),
+    )
+}
```
Comment (ryo33) [inline crates/parol/tests/parol_new.rs:68]: I've added snapshot tests for `parol new` command. One drawback is that it depends on the `diff` command.

File: crates/parol-ls/src/parol_ls_grammar_trait.rs (hunk 2 above)
Comment (ryo33) [reply crates/parol-ls/src/parol_ls_grammar_trait.rs]: Fixed

**Review by jsinger67 — APPROVED** [review]
Huge work done! Thank you 👍

**Comment by jsinger67** [comment]
Tests snapshot_lib and snapshot_bin are currently failing on Window OS.
Surely because `diff` is no native Windows command.
`Failed to spawn Command { cmd: "diff"...`
I think there is something like `comp`.
If you don't have the chance to fix this I can surely jump in.

**Comment by ryo33** [comment]
I’ll try `git diff`. From reading the git documentation, this usage seems to be expected. I expect `git` command is available on Windows.

**Comment by jsinger67** [comment]
Yes this could be fairly expected for people developing software 😉

**Pull request merged**